### ISSUE TITLE
advertisingIdentifier: reason for apple rejection.

### ISSUE THREAD

**Issue by yonkahlon** [BranchMetrics/xamarin-branch-deep-linking-attribution#134]
My app was rejected by Apple because I'm referencing IDFA

Their note:

> We noticed that your Kids Category app includes analytics, advertising and collects, transmits, or has the ability to share personal information or device information with third parties. Specifically:
> 
> - Your app includes third-party analytics or third-party advertising with the ability to collect, transmit or share identifiable information, including, for example, IDFA. We found that your app references the ASIdentifierManager API, which provides access to a user's IDFA, in the following location(s) in your binary:
> 
> • /System/Library/Frameworks/AdSupport.framework/AdSupport
> 
> It would be appropriate to remove all instances of “ASIdentifierManager” from your app, even if they are not utilized in your app's functionality.

I've seen this issue reported (and resolved?) on the other Branch repos:

https://github.com/BranchMetrics/react-native-branch-deep-linking-attribution/issues/601
https://github.com/BranchMetrics/ios-branch-deep-linking-attribution/issues/1052

Not sure where the reference is coming in - perhaps the static library (Branch.a) needs to be built with the BRANCH_EXCLUDE_IDFA_CODE.

I did try that, but if run `fgrep -R advertisingIdentifier .` I still find a few hits. But maybe I'm doing something wrong.

**Comment by echo-branch** [comment]
Ah I see. Xamarin uses a pre-compiled binary so you'll be unable to use the workaround that other platforms are using. I'll look into creating a custom nuget package with the option enabled.

**Comment by echo-branch** [comment]
Added a version of the nupkg to the releases page that has the no IDFA option enabled. We'll try to make it a regular part of our release process.

https://github.com/BranchMetrics/xamarin-branch-deep-linking-attribution/releases/tag/7.1.1

**Comment by yonkahlon** [comment]
Great, thanks for doing this so quickly! I'll submit my app for review and provide feedback

**Comment by yonkahlon** [comment]
Our app was rejected again, same reason. Looks like there is still a reference to ASIdentifierManager in the Branch.a file.

I did make sure that it compiled with the new nuget package.

**Comment by yonkahlon** [comment]
I've been trying go through the code, but I'm stuck at 2 points (I'm not so familiar with xcode):
 - Building the Universal Static library file: "Branch.a"
 - Building the nuget package or the 3x .dlls (Branch-Xamarin-Lib.iOS.dll, Branch-Xamarin-SDK.dll, Branch-Xamarin-SDK.iOS.dll)

**_Issue 1: Building "Branch.a"_**
When I run the "build_static_framework.sh" script, I get a .zip directory, but it only contains "Branch.xcframework". Inside that is the separated lib files for simulator and device.

I thought to try combine it with lipo -create but I got an error saying they both have the same architecture (arm64).

Could you provide some guidance so I can try build the .a file?


**_Issue 2: Building Xamarin Project_**
The wrapper project builds fine (Branch.Xamarin-Lib.iOS), but the SDK fails to build - it looks like the wrapper project is missing a lot of stuff. These are the error I get.

      Error CS0234: The type or namespace name 'BranchUniversalObject' does not exist in the namespace 'IOSNativeBranch' (are you missing an assembly reference?) (55, 73)

      Error CS0234: The type or namespace name 'BranchLinkProperties' does not exist in the namespace 'IOSNativeBranch' (are you missing an assembly reference?) (95, 73)

      Error CS0234: The type or namespace name 'Branch' does not exist in the namespace 'IOSNativeBranch' (are you missing an assembly reference?) (26, 27)

If I decompile the .dll it seems that nothing from the ApiDefinition.cs file is being linked. The .dll only contains the enums from Struct.cs

**Comment by echo-branch** [comment]
Now when I checked the Branch.a attached to the releases page, there was no references to advertisingIdentifier in the Branch.a.

No matches in the new file, but matches the file downloaded from nuget.
`
fgrep -R advertisingIdentifier Branch.a
`

I'm wondering if the package manager is downloading the one on nuget automatically during build?  Or possibly it's cached?  I find Visual Studio is rather aggressive about caching.

**Comment by echo-branch** [comment]
I tried enabling IDFA via the ATT prompt in my test app and don't see the IDFA in the Branch open payload.  That indicates the custom nuget package is working properly. I suspect your build is caching the default version and not using the custom one.

The process to verify is rather involved, but if you want to confirm yourself.

Enable logging, this should be called before the BranchIOS.Init
```
            // Enables logging, this is really an ugly way of doing it, need a cleaner API for end users
            IOSNativeBranch.Branch tmp = IOSNativeBranch.Branch.GetInstance("your_branch_key");
            tmp.EnableLogging();
```

Request permission for IDFA, this can be called whenever the app has started. Remember the device needs to allow requests for tracking in the privacy settings, or else you will always get denied.
```
        public void RequestATT()
        {
            AppTrackingTransparency.ATTrackingManager.RequestTrackingAuthorization((result) =>
            {
                switch (result)
                {
                    case AppTrackingTransparency.ATTrackingManagerAuthorizationStatus.NotDetermined:
                        break;
                    case AppTrackingTransparency.ATTrackingManagerAuthorizationStatus.Restricted:
                        break;
                    case AppTrackingTransparency.ATTrackingManagerAuthorizationStatus.Denied:
                        break;
                    case AppTrackingTransparency.ATTrackingManagerAuthorizationStatus.Authorized:
                        break;
                    default:
                        break;
                }
            });
            var status = AppTrackingTransparency.ATTrackingManager.TrackingAuthorizationStatus;
        }
```

Update the Info.plist to include NSUserTrackingUsageDescription. This can be any string.

Allow tracking via the prompt on a real device.

If you are using the default nuget package, you should see `hardware_id_type` set to idfa.

If you are using the custom nuget package, you should see `hardware_id_type` set to vendor_id.

**Comment by yonkahlon** [comment]
Thanks for all the info!

Yes, there is no reference to advertisingIdentifier, but I found a few to ASIdentifierManager (that's what Apple are complaining about). I'm not exactly sure where it's coming from. 

Yeah, I was worried about the nuget not updating to the correct one, and I verified in my build folder that it used the new .dll. I will redo the whole process and check if I made a mistake.

**Comment by yonkahlon** [comment]
It took me a long time to get there, but I couldn't build the binding myself [because of this visual studio issue](https://developercommunity2.visualstudio.com/t/Building-Xamarin-iOS-binding-project-not/1361154?entry=problem&ref=native&refTime=1620638930383&refUserId=a9579ba0-7d33-4f6d-9d1d-37465f4adcd3)

So I've compiled the iOS project with the BRANCH_EXCLUDE_IDFA_CODE macro defined, and I get a different result that's in your no-IDFA.zip file. Both mine and your build have **_no_** reference to advertisingIdentifier, but yours had a few references to ASIdentifierManager, while mine didn't. I've attached my Branch.a file.

`fgrep -R ASIdentifierManager Branch.a`

I'm wondering if your build folder wasn't clean at the time? It's strange for that to happen.

I'm going to submit again today with the .dll I generated and see if that works.

[Branch.zip](https://github.com/BranchMetrics/xamarin-branch-deep-linking-attribution/files/6452013/Branch.zip)
